### PR TITLE
Reopen 91719

### PR DESCRIPTION
**Pull request by shampoofactory**
Reopened #91719, which was closed inadvertently due to technical difficulties.

### CODE REVIEW

**Comment by workingjubilee** [comment]
r? @workingjubilee
Let's get the formalities out of the way as fast as possible.
@bors try @rust-timer queue

File: src/test/codegen/array-equality.rs [hunk 1]
```diff
@@ -47,11 +47,33 @@ pub fn array_eq_long(a: &[u16; 1234], b: &[u16; 1234]) -> bool {
...
+    // CHECK-NEXT: start:
+    // CHECK-NEXT: bitcast
+    // CHECK-NEXT: %[[LOAD:.+]] = load i128,
+    // CHECK-NEXT: %[[EQ:.+]] = icmp eq i128 %[[LOAD]], 0
     // CHECK-NEXT: ret i1 %[[EQ]]
     x == [0; 8]
 }
+
+// CHECK-LABEL: @array_eq_zero_long([1234 x i16]*
+#[no_mangle]
```
Comment (scottmcm) [inline src/test/codegen/array-equality.rs:71]: Sorry, I forget this when I suggested the codegen test.

**Comment by workingjubilee** [comment]
Excellent!

To [quote myself again:](https://github.com/rust-lang/rust/pull/91719#issuecomment-1051555624)
"That's even nicer than the last run. The highs aren't actually that much higher, but the dips are shallower and it's on average much better. It might be within random variance, but I suspect this actually neatly explains why what was once an optimization is now a regression: the Rust ABI was diverted down a path that LLVM did not focus much on optimizing in the next 5 versions, so as LLVM 14 is somewhat better at this than LLVM 13, likely LLVM 13 was better than LLVM 12, etc., all the way back to LLVM 9, when it was actually a win to take the other path."

@rustbot label: +perf-regression-triaged
@bors r+

**Pull request merged**